crates/sui-light-client/src/main.rs

```diff
@@ -93,12 +116,42 @@ struct Config {
 
     //  Genesis file name
     genesis_filename: PathBuf,
+
+    /// Object store url
+    object_store_url: String,
+
+    /// GraphQL endpoint
+    graphql_url: String,
 }
 
-impl Config {
-    pub fn rest_url(&self) -> String {
-        format!("{}/rest", self.full_node_url)
-    }
+async fn query_last_checkpoint_of_epoch(config: &Config, epoch_id: u64) -> anyhow::Result<u64> {
+    // GraphQL query to get the last checkpoint of an epoch
+    let query = json!({
+        "query": "query ($epochID: Int) { epoch(id: $epochID) { checkpoints(last: 1) { nodes { sequenceNumber } } } }",
+        "variables": { "epochID": epoch_id }
+    });
+
+    // Submit the query by POSTing to the GraphQL endpoint
+    let client = reqwest::Client::new();
+    let resp = client
+        .post(&config.graphql_url)
+        .header("Content-Type", "application/json")
+        .body(query.to_string())
+        .send()
+        .await
+        .expect("Cannot connect to graphql")
+        .text()
+        // .json::<HashMap<String, String>>()
+        .await
+        .expect("Cannot parse response");
+
+    // Parse the JSON response to get the last checkpoint of the epoch
+    let v: Value = serde_json::from_str(resp.as_str()).expect("Incorrect JSON response");
+    let checkpoint_number = v["data"]["epoch"]["checkpoints"]["nodes"][0]["sequenceNumber"]
```

Comment:
This may be a bit panick-y if one of these fields does not exist for some reason (like if the epoch was pruned). This is fine if we're intending this to be an example application, but if not, in some places we use `cynic` as a library to provide type-safe GraphQL queries in Rust -- where you will then be compelled to introduce `Option<...>` in the appropriate places where data might not exist, and check for it.

See this tool, for example, to fetch all packages:

https://github.com/MystenLabs/sui/blob/main/crates/sui-package-dump/src/query.rs

Reply:
Thanks for the great feedback, will look at `cynic` indeed.

Reply:
Feel free to also peek here: https://github.com/MystenLabs/sui-rust-sdk/tree/stefan/graphql_client_cynic/crates/sui-graphql-client
https://github.com/MystenLabs/sui-rust-sdk/tree/stefan/graphql_client_cynic/crates/sui-graphql-client#custom-queries